File: frontend/components/forms/dimensionUnits.ts
/**
 * CENTIMETRES ↔ INCHES FOR THE TOOL RECORD FORM, and the four rules that keep a live conversion from
 * eating the number it is converting.
 *
 * `ToolDocumentation` pairs `height` with `heightInches` and `width` with `breadthInches`. They were
 * two unrelated columns until this change — the form's own note said so and told researchers to fill
 * one and not the other — and they are now the SAME measurement in two units, filled from whichever
 * box the researcher types in. `lengthInches` has no centimetre partner and gains none.
 *
 * ── WHY THIS IS A MODULE AND NOT TWO EXPRESSIONS INSIDE THE FORM ────────────────────────────────
 * The same split, for the same reason, as `components/forms/measurementMethods.ts` and
 * `components/media/gridProposal.ts`: there is no React renderer in this repository's
 * devDependencies, so a judgement written inside JSX is exercised by nobody. What this file decides
 * is the class of thing whose broken state looks EXACTLY like its working state — a plausible number
 * in a box — until somebody reads the record a year later. `e2e/dimension-units-unit.spec.ts` drives
 * every row of the table this was specified against.
 *
 * ── THE SAME ARITHMETIC RUNS ON THE HANDSET, AND THE TWO MUST NOT DRIFT ─────────────────────────
 * `android/app/src/main/java/com/fieldrepository/app/ui/DimensionUnits.kt` is the Kotlin twin, with
 * the same four function names and the same worked table in its own JVM test. Kotlin's
 * `String.toDouble` and JavaScript's `Number()` disagree at the edges — `Number("0x1A")` is 26 while
 * `"0x1A".toDouble()` throws, `"1.5f".toDouble()` is 1.5 while `Number("1.5f")` is `NaN`, and both
 * accept `"Infinity"` and `"NaN"` — so the grammar below is a regular expression both languages
 * implement identically rather than either language's own parser. If you change a rule here, change
 * it there.
 *
 * ── AND IT IS **NOT** THE BACKEND'S `_inches_to_cm` ─────────────────────────────────────────────
 * The sibling application's `design_workshops._inches_to_cm` is `round(float(v) * 2.54, 2)` —
 * Python's banker's rounding, which breaks a halfway value TOWARDS THE EVEN hundredth while this
 * file breaks it upwards. The two live on opposite sides of the wire, are applied to different
 * columns, and must not be "unified": a single shared helper would silently move every figure that
 * lands on a halfway value in whichever of the two it was not written for. This repository has no
 * design-workshop carry at all, so there is nothing here to unify them WITH — the paragraph is kept
 * so that the next reader porting a change from the sibling repository does not try.
 */

/**
 * EXACT, by definition rather than by measurement: the inch has been 25.4 mm since the 1959
 * international yard-and-pound agreement. Never "2.5 for readability", and never a pre-divided
 * `0.3937…` — see the association-order note on the two converters below.
 */
export const CM_PER_INCH = 2.54;

/**
 * The largest value `Decimal(10, 2)` can hold, in hundredths: 99,999,999.99.
 *
 * A conversion that overflows the column writes NOTHING rather than a truncated number. Postgres
 * refuses the value at the boundary with `numeric field overflow`, which on a save reaches the
 * researcher as an opaque 500 — long after the box was filled in, and naming a column rather than
 * the box they typed into.
 */
const DECIMAL_10_2_MAX_SCALED = 9999999999;

/**
 * What counts as a number in a dimension box, written as one grammar for both clients.
 *
 * Deliberately NARROWER than either language's own string-to-double: no sign (a negative dimension
 * is not a measurement, and every one of these boxes carries `min={0}`), no exponent, no hexadecimal,
 * no `Infinity`, no `NaN`, no thousands separator. Leading and trailing spaces and tabs are
 * tolerated because a paste carries them.
 *
 * `"1."` AND `".5"` ARE ADMITTED ON PURPOSE. They are what a person typing a decimal produces
 * halfway through, and the consequence is spelled out at {@link propagate}.
 */
const NUMERIC_RE = /^[ \t]*([0-9]+\.?[0-9]*|\.[0-9]+)[ \t]*$/;

/**
 * The number in a dimension box, or `null` when the box does not hold one.
 *
 * A total function: everything the grammar admits is finite and non-negative by construction, and
 * the `Number.isFinite` line below is kept anyway so that the type says so without the reader having
 * to re-derive it from the regular expression.
 */
export function parseDimension(text: string): number | null {
  if (!NUMERIC_RE.test(text)) return null;
  const value = Number(text.trim());
  return Number.isFinite(value) ? value : null;
}

/**
 * A hundredths-scaled integer rendered as the shortest honest decimal string.
 *
 * BUILT FROM THE INTEGER, NEVER FROM `Number.prototype.toString`. JavaScript prints `"3"` for an
 * integral `3.0` where Kotlin prints `"3.0"`, and the two disagree again on large magnitudes — so
 * the only way four clients can agree on the text in the box is to never let either language's
 * float formatter near it.
 */
function renderScaled(scaled: number): string {
  const whole = Math.floor(scaled / 100);
  const frac = scaled % 100;
  if (frac === 0) return String(whole);
  if (frac % 10 === 0) return `${whole}.${frac / 10}`;
  return `${whole}.${String(frac).padStart(2, "0")}`;
}

/**
 * Round half UP (towards +∞), in one expression, on a value the grammar guarantees is non-negative.
 *
 * `Math.round` is not used, and the reason is not style: `kotlin.math.round` is half-AWAY-FROM-ZERO
 * while the JVM's `Math.round` is `floor(x + 0.5)`, so the twin implementations would agree on every
 * value this form can produce and disagree the first time somebody reuses one of them on a signed
 * quantity. `floor(x + 0.5)` is the rule written down rather than inherited.
 *
 * ONE ROUNDING, NEVER TWO. Rounding to two decimals and then formatting from a re-derived double
 * rounds twice and can move the last hundredth.
 */
function scaleHalfUp(value: number): number | null {
  const scaled = Math.floor(value + 0.5);
  return scaled > DECIMAL_10_2_MAX_SCALED ? null : scaled;
}

/**
 * The centimetre text for an inch box's contents, or `null` when there is nothing to write.
 *
 * `value * CM_PER_INCH * 100` IS LEFT-TO-RIGHT AND MUST STAY SO. Rewriting it as `value * 254` is
 * not the same computation in binary64: each rewrite moves the last ulp and can flip a hundredth.
 * The worked table in `e2e/dimension-units-unit.spec.ts` is what catches that, and `1.25 → 3.18`
 * (whose intermediate is exactly 317.5) is the row that catches it first.
 */
export function cmTextFromInches(text: string): string | null {
  const inches = parseDimension(text);
  if (inches === null) return null;
  const scaled = scaleHalfUp(inches * CM_PER_INCH * 100);
  return scaled === null ? null : renderScaled(scaled);
}

/** The inch text for a centimetre box's contents, or `null`. Same association rule as above. */
export function inchesTextFromCm(text: string): string | null {
  const cm = parseDimension(text);
  if (cm === null) return null;
  const scaled = scaleHalfUp((cm / CM_PER_INCH) * 100);
  return scaled === null ? null : renderScaled(scaled);
}

/**
 * Write the partner box for one keystroke in the box the researcher is actually typing in.
 *
 * ── THE THREE RULES, IN THE ORDER THEY ARE TESTED ───────────────────────────────────────────────
 *
 * 1. **AN EMPTY BOX CLEARS ITS PARTNER**, and this branch comes BEFORE the parse because it is the
 *    one input that is not a number and still means something: empty means *no value*, and leaving a
 *    stale converted number standing beside it is a lie the record cannot detect afterwards.
 * 2. **ANYTHING ELSE THAT IS NOT A NUMBER WRITES NOTHING.** `"abc"`, `"1.2.3"`, `"1e3"` and an
 *    overflow of `Decimal(10, 2)` all leave the partner exactly as it is.
 * 3. **A PARTIALLY TYPED DECIMAL IS A NUMBER, NOT A MISTAKE:** `1.` parses and converts, so the
 *    partner tracks the keystrokes instead of blanking and refilling. Only a string that cannot be a
 *    number at all leaves the partner alone — and only an EMPTY box clears it, because empty is the
 *    one input that means "no value". Typing `1` → `1.` → `1.5` walks the partner `2.54` → `2.54` →
 *    `3.81`; typing `1.2` → `1.2.` holds `3.05` for one keystroke and corrects on the next, which is
 *    strictly better than a partner that blanks, because a blank partner is rule 1's statement and
 *    somebody mid-decimal has not made it.
 *
 * ── IT IS CALLED FROM THE SOURCE BOX'S OWN `onChange` AND FROM NOWHERE ELSE ─────────────────────
 * One-directional per keystroke: the box being typed in writes its partner, and the partner never
 * writes back. A `useEffect` watching both values is the shape that cannot work — it fires for
 * whichever of the two changed, so 1 cm becomes 0.39 in becomes 0.99 cm and a round trip eats the
 * value. React's `setState` does not re-invoke the target input's `onChange`, so the natural shape
 * is already the correct one; it is written down here because the effect version reads like a
 * tidy-up.
 *
 * ── AND IT NEVER RUNS ON LOAD ──────────────────────────────────────────────────────────────────
 * Historic rows genuinely hold unrelated numbers in `height` and `heightInches` — that is what the
 * old on-screen note was about — so a conversion at mount would rewrite a saved value the moment
 * somebody merely OPENED the record. Each box is seeded from its own column and nothing else.
 */
export function propagate(
  text: string,
  convert: (text: string) => string | null,
  setPartner: (value: string) => void
): void {
  if (text.trim() === "") {
    setPartner("");
    return;
  }
  const next = convert(text);
  if (next !== null) setPartner(next);
}

File: android/app/src/main/java/com/fieldrepository/app/ui/DimensionUnits.kt
package com.fieldrepository.app.ui

import kotlin.math.floor

/**
 * CENTIMETRES ↔ INCHES FOR THE TOOL FORM'S PAIRED DIMENSION BOXES.
 *
 * The web twin of this file is `frontend/components/forms/dimensionUnits.ts`, and the two were
 * written together on purpose — the same reason `ui/RecordPickers.kt` names its own twin in its
 * header. Four clients across two repositories convert the same two boxes, and the failure mode of
 * one rule implemented four times is that they agree on the day they are written and disagree on a
 * halfway value six months later, in a column nobody re-reads.
 *
 * IN A `ui/` FILE RATHER THAN IN `MainActivity.kt` BECAUSE THAT IS WHERE A JVM UNIT TEST CAN REACH
 * IT — exactly the reason `ui/RecordPickers.kt`'s header gives for `craftChangeClearsArtisan`
 * living there. `DimensionUnitsTest` drives every row of the table the four clients share.
 *
 * ── WHAT IS PAIRED, AND WHAT DELIBERATELY IS NOT ────────────────────────────────────────────────
 *
 *      Height (cm)  `height`  ↔  Height (inches)  `heightInches`
 *      Width  (cm)  `width`   ↔  Breadth (inches) `breadthInches`
 *      Length (inches) `lengthInches` — STANDALONE. No centimetre partner, no new column.
 *
 * No column is added, renamed or retyped by this and no wire key changes: `height` and `width` keep
 * their names everywhere, and only what a person READS on the label changed. What DID change is the
 * old claim beside those two boxes — that `height` is "the unit-less legacy column" and must not be
 * filled alongside `heightInches`. It is the centimetre box now, the two are one measurement in two
 * units, and filling either fills the other.
 *
 * ── THE ARITHMETIC, AND WHY IT IS SPELLED THIS PEDANTICALLY ─────────────────────────────────────
 *
 * `2.54` is EXACT: the inch has been defined as 25.4 mm since the 1959 international yard-and-pound
 * agreement. Never "2.5 for readability".
 *
 * Parsing goes through ONE grammar rather than through each language's own number reader, because
 * `Double.parseDouble` and JavaScript's `Number()` disagree at the edges: `Number("0x1A")` is 26
 * while `"0x1A".toDouble()` throws, `"1.5f".toDouble()` is 1.5 while `Number("1.5f")` is NaN, and
 * both accept `"Infinity"` and `"NaN"`. A regex both languages implement identically is the only
 * shape in which the four clients cannot come apart.
 *
 * Conversion and rounding are ONE step, producing a SCALED INTEGER — hundredths, because both
 * columns are `Decimal(10,2)`. Rounding to two places and then formatting from a re-derived double
 * would round twice and can move the last hundredth. Rendering is built from that integer and never
 * from `Double.toString()`: Kotlin prints `3.0` where JavaScript prints `3`, and the two disagree
 * again on large magnitudes.
 *
 * `floor(x + 0.5)` and not `kotlin.math.round`, which is half-away-from-zero, nor `Math.round`,
 * which is `floor(x + 0.5)` only by its own contract rather than by this file's. The values here
 * are never negative — the grammar admits no sign — so all three rules agree today; writing the
 * expression out is what keeps them agreeing if a signed value ever reaches this file.
 *
 * ASSOCIATION ORDER IS LOAD-BEARING. `value * CM_PER_INCH * 100.0` must NOT be rewritten as
 * `value * 254.0`, and `value / CM_PER_INCH * 100.0` must NOT be rewritten as
 * `value * (100.0 / CM_PER_INCH)` or as a precomputed `0.3937007874015748`. Each rewrite changes the
 * last ulp and can flip a hundredth — 1.25 inches is 317.5 hundredths of a centimetre EXACTLY under
 * this spelling, and the halfway rule then has something true to round.
 *
 * ── NOT THE SAME FUNCTION AS THE SERVER'S ───────────────────────────────────────────────────────
 *
 * The sibling repository's `backend/app/services/design_workshops._inches_to_cm` is
 * `round(float(v) * 2.54, 2)` — Python's banker's rounding, a DIFFERENT halfway rule from
 * `floor(x + 0.5)`. It lives on the other side of the wire, it is applied to a different column (a
 * design-workshop stage carry, not a record box), and the two must not be "unified": doing so would
 * change every carried figure that lands on a halfway value. This repository has no design
 * workshops and therefore no copy of that function at all; the paragraph is here so that the next
 * reader who meets it in the sibling does not fold one into the other.
 */

/** Centimetres in one inch. Exact by definition since 1959; see the file header. */
const val CM_PER_INCH: Double = 2.54

/**
 * The one grammar both clients parse a dimension box with.
 *
 * Leading and trailing spaces or tabs, then either digits with an optional fractional part or a
 * bare fractional part. NO SIGN, no exponent, no hex, no thousands separator, no trailing type
 * letter. `"1."` and `".5"` ARE admitted, deliberately: they are what a person mid-decimal has
 * typed, and see [propagateDimension] for why that matters.
 */
private val NUMERIC_RE = Regex("""^[ \t]*([0-9]+\.?[0-9]*|\.[0-9]+)[ \t]*$""")

/** 99,999,999.99 in hundredths — the ceiling of the `Decimal(10,2)` columns on both sides. */
private const val DECIMAL_10_2_MAX_SCALED: Long = 9_999_999_999L

/**
 * [text] as a finite, non-negative number, or null when it is not one at all.
 *
 * A TOTAL FUNCTION: every string has an answer and none of them is an exception. The `isFinite`
 * guard is unreachable under the grammar above and is kept anyway, because "unreachable" is a
 * property of the regex and the regex is one edit away from somebody allowing an exponent.
 */
fun parseDimension(text: String): Double? {
    if (!NUMERIC_RE.matches(text)) return null
    val value = text.trim().toDoubleOrNull() ?: return null
    if (!value.isFinite()) return null
    return value
}

/** [text] read as inches, in hundredths of a centimetre — or null when it is not a number. */
private fun cmScaledFromInches(text: String): Long? {
    val value = parseDimension(text) ?: return null
    val n = floor(value * CM_PER_INCH * 100.0 + 0.5)
    if (n > DECIMAL_10_2_MAX_SCALED) return null
    return n.toLong()
}

/** [text] read as centimetres, in hundredths of an inch — or null when it is not a number. */
private fun inchesScaledFromCm(text: String): Long? {
    val value = parseDimension(text) ?: return null
    val n = floor(value / CM_PER_INCH * 100.0 + 0.5)
    if (n > DECIMAL_10_2_MAX_SCALED) return null
    return n.toLong()
}

/**
 * A scaled hundredth count as the shortest decimal string that means it.
 *
 * 254 gives "254", 2540 gives "25.4", 318 gives "3.18", 4 gives "0.04". Built from the integer and
 * never from a double — see the file header.
 */
private fun renderScaled(n: Long): String {
    val whole = n / 100
    val frac = n % 100
    return when {
        frac == 0L -> whole.toString()
        frac % 10L == 0L -> "$whole.${frac / 10}"
        else -> "$whole." + frac.toString().padStart(2, '0')
    }
}

/** An inches box's text as the centimetre partner's text, or null to write nothing. */
fun cmTextFromInches(text: String): String? = cmScaledFromInches(text)?.let(::renderScaled)

/** A centimetre box's text as the inches partner's text, or null to write nothing. */
fun inchesTextFromCm(text: String): String? = inchesScaledFromCm(text)?.let(::renderScaled)

/**
 * Write the partner of the box a person is TYPING IN — the whole of the propagation rule, in one
 * place, so the four boxes on the tool form cannot each implement a slightly different one.
 *
 * ONE-DIRECTIONAL, ALWAYS. Only the box being edited writes its partner; the partner never writes
 * back. That is achieved by calling this from inside the source box's own `onValueChange` and by
 * NOTHING ELSE — a `LaunchedEffect(height, heightInches)` watcher fires for BOTH values and is how
 * 1 cm becomes 0.39 in becomes 0.99 cm. Compose state assignment does not re-invoke the target
 * control's change handler, so the natural shape is already the correct one.
 *
 * AND IT COVERS DICTATION FOR FREE, WHICH IS NOT OBVIOUS. `ui/RecordProseField.kt` writes a box
 * from two places — the keyboard's `onValueChange` on the `OutlinedTextField`, and the recogniser's
 * `onCommit` — but BOTH of them call the CALLER's own `onValueChange` lambda, so a rule written
 * there runs for a spoken value exactly as for a typed one. A rule written inside the
 * `OutlinedTextField` instead would miss every dictated value, silently. It is the same reason
 * `markers.forget(...)` already sits in those lambdas rather than in the control.
 *
 * NEVER ON LOAD. Nothing here is reachable from form construction: each box is seeded from its own
 * column and from nothing else. Historic rows genuinely hold unrelated numbers in `height` and
 * `heightInches`, and a load-time conversion would destroy one of them on the next save.
 *
 * CLEARING PROPAGATES, and it is the ONE case where a non-numeric value writes the partner: an
 * empty box means "no value", and leaving a stale converted number beside it is a lie. It is spelled
 * as its own branch BEFORE the parse for exactly that reason.
 *
 * A PARTIALLY TYPED DECIMAL IS A NUMBER, NOT A MISTAKE: "1." parses and converts, so the partner
 * tracks the keystrokes instead of blanking and refilling. Only a string that cannot be a number at
 * all leaves the partner alone — and only an EMPTY box clears it, because empty is the one input
 * that means "no value". So 1 → 1. → 1.5 gives 2.54 → 2.54 → 3.81, and 1.2 → 1.2. holds 3.05 for
 * one keystroke and corrects on the next, which is strictly better than a partner that blanks in
 * the middle of a number the researcher has not finished typing.
 */
fun propagateDimension(text: String, convert: (String) -> String?, setPartner: (String) -> Unit) {
    if (text.trim().isEmpty()) {
        setPartner("")
        return
    }
    val next = convert(text) ?: return
    setPartner(next)
}
